### PR TITLE
Add Review and Code Style guidelines, Lint rules and more

### LINKED ISSUE #22: Add Swiftlint, review guidelines
As a Contributor I'd like to know what the code style and the review guidelines are, so that I can be more effective with my PRs

- [x] Add SwiftLint guide from https://github.com/raywenderlich/swift-style-guide/blob/master/
- [x] Add SwiftLint yml
- [x] Add Code Review Guidelines
- [x] add Pull request template that contains a checklist for both contributor(s) and reviewer(s)

Resources:
https://swift.org/documentation/api-design-guidelines/
https://google.github.io/eng-practices/review/reviewer/
https://zcash.readthedocs.io/en/latest/rtd_pages/development_guidelines.html

### PR DESCRIPTION
**Pull request by pacu**
closes #22 

This PR intends to answer two basic questions:
- How should contributors structure their Pull Request?
- How should PR be reviewed?

Things included:
- Add SwiftLint guide from https://github.com/raywenderlich/swift-style-guide/blob/master/
- Add SwiftLint yml
- Add Code Review Guidelines
- Add Pull request template that contains a checklist for both contributor(s) and reviewer(s)
- Add Bug report, feature request and UX request templates based on Zcashd's templates

### CODE REVIEW

File: co.electriccoin.swiftlint.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  - toggle_bool
+  # - trailing_closure # weird in SwiftUI
+  - unneeded_parentheses_in_closure_argument
+  - unused_import
+  - vertical_whitespace_closing_braces
+  - vertical_whitespace_opening_braces
+  - yoda_condition
+
+
+custom_rules:
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:59]: A few more custom rules to consider adding:

string_concatenation:
    included: ".*\\.swift"
    excluded: ".*Test\\.swift"
    name: "String Concatenation"
    regex:  " \\+ \"|\" \\+ |\\+= \""
    message: "Please use string interpolation instead of concatenation"
    severity: error
    
    
print_function_usage:
    included: ".*\\.swift"
    excluded: ".*Test\\.swift"
    name: "Swift `print()` or `debugPrint()` should not be used in App Code"
    regex: "print\\(|debugPrint\\("
    message: "The Swift `print()` or `debugPrint()` functions should not be used."
    severity: warning

nslog_function_usage:
    included: ".*\\.swift"
    excluded: ".*Test\\.swift"
    name: "Swift `NSLog()` should not be used in App Code"
    regex: "NSLog\\("
    message: "The swift `NSLog` function should not be used."
    severity: error

File: co.electriccoin.swiftlint.yml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    - i
+    - id
+    - x
+    - y
+    - z
+
+indentation_width:
+  indentation_width: 2
+
+line_length:
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:93]: Should we also include an actual line length value here, like 150 ?

File: co.electriccoin.swiftlint.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  ignores_urls: true
+  ignores_function_declarations: true
+  ignores_comments: true
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:96]: I would recommend we also add a file length here.  A warning at 600 lines and an error at 1000 seems like a reasonable place to start.

File: co.electriccoin.swiftlint.yml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  - ${PWD}/Pods
+  - ${PWD}/DerivedData
+  - ${PWD}/xctemplates
+
+disabled_rules:
+  - discarded_notification_center_observer
+  - notification_center_detachment
+  - orphaned_doc_comment
+  - todo
+  - unused_capture_list
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:15]: Some more rules to consider disabling by default
-nesting (allow for types to be nested, common pattern in Swift)
-multiple_closures_with_trailing_closure
- generic_type_name (allow for arbitrarily long generic type names)

File: co.electriccoin.swiftlint.yml [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  - enum_case_associated_values_count
+  - fatal_error_message
+  - first_where
+  - force_unwrapping
+  - implicitly_unwrapped_optional
+  - indentation_width
+  - last_where
+  - legacy_random
+  - literal_expression_end_indentation
+  - multiline_arguments
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:38]: multiline_argument_brackets should probably be added here

File: co.electriccoin.swiftlint.yml [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  - multiline_function_chains
+  - multiline_literal_brackets
+  - multiline_parameters
+  - multiline_parameters_brackets
+  - operator_usage_whitespace
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:43]: no_space_in_method_call should probably be added here

File: co.electriccoin.swiftlint.yml [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  - overridden_super_call
+  - pattern_matching_keywords
+  - prefer_self_type_over_type_of_self
+  - redundant_nil_coalescing
+  - redundant_type_annotation
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:48]: return_arrow_whitespace should probably be added here

File: co.electriccoin.swiftlint.yml [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  - strict_fileprivate
+  - toggle_bool
+  # - trailing_closure # weird in SwiftUI
+  - unneeded_parentheses_in_closure_argument
+  - unused_import
+  - vertical_whitespace_closing_braces
+  - vertical_whitespace_opening_braces
```
Comment (adamstener) [inline co.electriccoin.swiftlint.yml:55]: weak_delegate has proven very useful in the past, it should probably be added here.

File: co.electriccoin.swiftlint.yml (hunk 5 above)
Comment (adamstener) [reply co.electriccoin.swiftlint.yml:38]: `mark` can be added here to add some helpful hints on whether or not your pragma marks are syntactically correct and provide the correct segmentation in the file hierarchy browser.

File: CODE_REVIEW_GUIDELINES.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+- Does every change make sense? Make sure to ask questions, even or especially "basic" coding questions ("what does this mean in Swift?") and domain-newbie questions ("what is the priority system, and why does this line change the priority?").
+
+In summary, here's a checklist that summarizes what reviewers should be looking for when doing a Code Review [1]
+
+- The code is well-designed.
+- The functionality is good for the users of the code.
+- Any UI changes are sensible and look good.
+- Any parallel programming is done safely.
+- The code isn’t more complex than it needs to be.
+- The developer isn’t implementing things they might need in the future but don’t know they need now.
```
Comment (adamstener) [inline CODE_REVIEW_GUIDELINES.md:38]: The developer isn’t implementing things they might need in the future but don’t know they need now or aren't using now.  - What do you think about this small addition at the end?

File: CODE_REVIEW_GUIDELINES.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+- Code has appropriate unit tests.
+- Tests are well-designed.
+- The developer used clear names for everything.
+- Comments are clear and useful, and mostly explain why instead of what.
+- Code is appropriately documented (generally in g3doc).
+- The code conforms to our style guides.
+- Make sure to review every line of code you’ve been asked to review, look at the context, make sure you’re improving code health, and compliment developers on good things that they do.
+
```
Comment (adamstener) [inline CODE_REVIEW_GUIDELINES.md:46]: I'm not sure where this would go, but do you think that this PR would benefit from a section that would discuss platform specific best practices?  I.e. are we going to prefer SwiftUI and discourage InterfaceBuilder and UIKit?  When using SwiftUI, should we document the discouraged usage of AnyView whenever possible?

**Review by adamstener — COMMENTED** [review]
This looks really good, I really like all the templates that were added and how much detail they have for reporting or opening an issue on GitHub.  Great job pulling all of this together, I just have a few comments or questions 👍

File: .github/pull_request_template.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), etc.)
+- [ ] Run the app: Did you run the app and try the changes? 
+- [ ] Did you provide Screenshots of what the App looks like before and your changes? (only applicable to UI Changes)
+- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.
+
+
+# Reviewer
+
+- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
+- [ ] Ad hoc review: Did you perform an ad hoc review?
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply .github/pull_request_template.md:17]: More or less, it should mean: "In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass."

That's probably wordy, which is why I originally used "ad hoc".

One thing I have done to make such explanations visually less "heavy" is use italics to represent the explanation.  Example:

- [ ] Ad hoc review: Did you perform an ad hoc review? _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._

File: .github/pull_request_template.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
+This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).
+
+# Author
+<!-- NOTE: Do not modify these when initially opening the pull request.  This is a checklist template that you tick off AFTER the pull request is created. -->
+- [ ] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
+- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
+- [ ] Code coverage: Did you check the code coverage report for the automated tests? Coverage can be generated by running `./gradlew check -PisCoverageEnabled=true; ./gradlew jacocoTestReport -PisCoverageEnabled=true` or by downloading the test report from the CI results. While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
```
[the remark this answers is not in the document]
Comment (pacu) [reply .github/pull_request_template.md:7]: I think we can remove the coverage thing until we settle on how to we want it to be. Have you worked with coverage and SwiftUI? how does it affect metrics ?

[the remark this answers is not in the document]
Comment (adamstener) [reply .github/pull_request_template.md:7]: Ideally you would not unit test your UI layer, and instead, remove any business logic from the view itself and put it in a more appropriate ViewModel or something where it can be unit tested.  The more you can keep your UI layer as simply a presentation layer the better.  There will be some logic in your views, like conditionally showing one view or another, and this could eventually be Snapshot tested, but to start I would suggest focusing on moving as much logic as you can into a layer that can actually be unit tested well and worry about the UI portion at a later date when or if it is needed.

File: CODE_REVIEW_GUIDELINES.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+#  Specification:
+
+- Are the requirements for the change well specified? Where are they documented? Bugfixes need a clear specification of the bug's cause and fix. (Small PRs might specify in a github ticket. Large changes may require stand-alone docs, or ZIPs.)
+- Is there a documented test plan, which describes how to manually verify the change works on testnet prior to a release?
+
+# User Documentation:
+
+- What is the "changelog" entry for this change? (All changes should include this.)
```
[the remark this answers is not in the document]
Comment (pacu) [reply CODE_REVIEW_GUIDELINES.md:14]: nice catch.

File: co.electriccoin.swiftlint.yml [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+# https://github.com/raywenderlich/swift-style-guide
+
+excluded:
+  - ${PWD}/Carthage
+  - ${PWD}/Pods
+  - ${PWD}/DerivedData
+  - ${PWD}/xctemplates
+
+disabled_rules:
+  - discarded_notification_center_observer
```
[the remark this answers is not in the document]
Comment (pacu) [reply co.electriccoin.swiftlint.yml:11]: I don't recall adding it. If you think is useful not to discard it, I'll remove it!

File: co.electriccoin.swiftlint.yml [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  
+private_over_fileprivate:
+  validate_extensions: true
+
+trailing_whitespace:
+  ignores_empty_lines: true
+  ignores_comments: true
+
+vertical_whitespace:
+  max_empty_lines: 2
```
[the remark this answers is not in the document]
Comment (pacu) [reply co.electriccoin.swiftlint.yml:110]: If I recall correctly, I added used 2 because it counts comments as empty lines. I don't like code starting right after a `// MARK:` comment.  

Maybe there's another way to sort that out?

[the remark this answers is not in the document]
Comment (adamstener) [reply co.electriccoin.swiftlint.yml:110]: It should not count comments as empty lines.  I agree with you about having a space after the MARK and that should be allowed with max_empty_lines: 1 I believe.  Maybe it's because of `ignores_comments: true` above?  If so we could remove that one.

File: co.electriccoin.swiftlint.yml (hunk 3 above)
Comment (pacu) [reply co.electriccoin.swiftlint.yml:96]: let's start with a warning

File: CODE_REVIEW_GUIDELINES.md (hunk 10 above)
Comment (pacu) [reply CODE_REVIEW_GUIDELINES.md:46]: probably on a platform specific section in the docs/ folder

Comment (adamstener) [reply CODE_REVIEW_GUIDELINES.md:47]: That seems appropriate for the scope of this PR being that it doesn't need to be exhaustive from the beginning, just some high level comments.

File: co.electriccoin.swiftlint.yml (hunk 2 above)
Comment (adamstener) [reply co.electriccoin.swiftlint.yml:121]: It looks like a warning was added to the line length that is more appropriate for a new section on "file length".  Can we change the 600 below to something like 150 and add a new section on file_length that can have the warning at 600?

File: secant.xcodeproj/project.pbxproj [hunk 16]
```diff
@@ -12,6 +12,8 @@
 		0D1922ED26BDE0C600052649 /* AppRouter.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0D1922EC26BDE0C600052649 /* AppRouter.swift */; };
 		0D1922F226BDE29300052649 /* ZcashSDKStubs.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0D1922F126BDE29300052649 /* ZcashSDKStubs.swift */; };
 		0D1922F826BDEB3500052649 /* MockServices.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0D1922F726BDEB3500052649 /* MockServices.swift */; };
+		0D24267726E9727C00DA2A47 /* WelcomeScreenTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0D24267626E9727C00DA2A47 /* WelcomeScreenTests.swift */; };
```
Comment (adamstener) [inline secant.xcodeproj/project.pbxproj:15]: Just an FYI it looks like there was some accidentally added code here.

File: .github/pull_request_template.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ccjernigan) [reply .github/pull_request_template.md]: This step should just be about getting authors to run the coverage tool and look at the output.  So this could be less about a metric and more about encouraging use of a tool to make important test omissions obvious.

It will usually make it very obvious if an important method was missed in terms of coverage, and then the author will be motivated to improve it.

Should using automation to block merging be a different discussion?  For example, to @adamstener 's points out that view code might have less coverage.  That would negatively impact metrics, but perhaps the metric isn't the thing we want to optimize for.

[the remark this answers is not in the document]
Comment (adamstener) [reply .github/pull_request_template.md]: I would imagine that any code coverage automation you have could ignore files or folders of UI code so that metrics aren't skewed.

File: .github/pull_request_template.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).
+
+# Author
+<!-- NOTE: Do not modify these when initially opening the pull request.  This is a checklist template that you tick off AFTER the pull request is created. -->
+- [ ] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
+- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
+- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
+- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), etc.)
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply .github/pull_request_template.md:8]: One pain point with GitHub templates—the hyperlinks here will point to the main branch and not the branch under review.  I haven't ever found a variable that allows the PR to refer to the current branch, but if such a variable existed that would be what we really want.

I mention this, because reviewers might click the link and see the version on `main` but not realize that's what they're viewing.

What kinds of ways could this be made less likely to trip people up?

File: .github/pull_request_template.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+- [ ] Run the app: Did you run the app and try the changes? 
+- [ ] Did you provide Screenshots of what the App looks like before and your changes? (only applicable to UI Changes)
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply .github/pull_request_template.md:10]: Should we clarify where the screenshots should go?

E.g. we probably want them added as a comment to the pull request, right?  (Alternative is that someone might commit them to the repo?)

File: .github/pull_request_template.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (adamstener) [reply .github/pull_request_template.md]: Until we have a tool for this though, I agree with @pacu that it should be removed, it will probably only add confusion at this point to reference a tooling for android that is not present in this repo.

File: .github/pull_request_template.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.
+
+
+# Reviewer
+
+- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
+- [ ] Ad hoc review: Did you perform an ad hoc review? _Did you perform an ad hoc review? In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
+- [ ] Automated tests: Did you review the automated tests?
+- [ ] Manual tests: Did you review the manual tests?
```
Comment (ccjernigan) [inline .github/pull_request_template.md:19]: There's no corresponding element for the author about writing manual tests.

If we decide we want to encourage drafting manual test plans, the author section might include:

- [ ] Manual tests: Did you update the manual tests under `/docs/tests`) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down and store them under `/docs/tests` _

File: CODE_REVIEW_GUIDELINES.md [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+- Are there any changes which require updates to user-facing documentation? If so, does the new documentation make sense? (documentation should be placed in [docs/](/docs) folder of the repo)
```
Comment (ccjernigan) [inline CODE_REVIEW_GUIDELINES.md:15]: user-facing or developer facing?

File: CODE_REVIEW_GUIDELINES.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+# Testing:
+
+- For non-minor PRs (up to the code reviewers and PR creator if this is needed), we require authors to perform a thorough self-review and self-test of the resulting code base, including use cases that might not be visible affected by the introduced changes. Reviewers are not expected to run the changes locally, but are definitely encouraged to do so at their best judgement.
```
Comment (ccjernigan) [inline CODE_REVIEW_GUIDELINES.md:19]: Typo: "be visible affected" -> "be visibly affected"

File: CODE_REVIEW_GUIDELINES.md [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+- Have all of the automated tests been executed by CI? Read over the CI report to verify this. (Note: our CI system may not currently provide test results prior to review. We should change this! In the meantime, the reviewer should run the tests.)
+- Is it clear that the test plan does actually test the new change? Does the test plan include any edge case / negative case / error conditions?
+
+# Code Review:
+- Does the code structure make sense? Does it follow existing conventions / frameworks of existing code, or is it introducing new abstractions / structure?
+- Is the code style consistent with the [Swift Style Guide](SWIFTLINT.md)?
+- Does every change make sense? Make sure to ask questions, even or especially "basic" coding questions ("what does this mean in Swift?") and domain-newbie questions ("what is the priority system, and why does this line change the priority?").
+
+In summary, here's a checklist that summarizes what reviewers should be looking for when doing a Code Review [1]
```
Comment (ccjernigan) [inline CODE_REVIEW_GUIDELINES.md:32]: Is this an ordered or unordered list?

Would an ordered list help in terms of clarifying conflicting priorities?

File: SWIFTLINT.md [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+	* [Force unwrapping](#force-unwrapping)
+	* [SwiftUI and multiple trailing closures](#swiftui-and-multiple-trailing-closures)
+	* [Open-source files](#open-source-files)
+* [Other notes](#other-notes)
+
+## Installing SwiftLint
+
+
+```bash
+brew install swiftlint
```
Comment (ccjernigan) [inline SWIFTLINT.md:30]: Can we add a line for installing Homebrew?  If someone doesn't have it installed yet, then this command will fail.  I see the note about Homebrew below, but someone following the instructions from top to bottom might miss that.

File: SWIFTLINT.md [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+	* [Force cast](#force-cast)
+	* [Force unwrapping](#force-unwrapping)
+	* [SwiftUI and multiple trailing closures](#swiftui-and-multiple-trailing-closures)
+	* [Open-source files](#open-source-files)
+* [Other notes](#other-notes)
+
+## Installing SwiftLint
+
+
+```bash
```
Comment (ccjernigan) [inline SWIFTLINT.md:29]: I never knew of this syntax highlighting feature until just now 🤯.

Very minor but does GitHub support `zsh`?  Although it's basically the same as `bash` for most usages, macOS switched to zsh as the default as of Catalina.

File: SWIFTLINT.md [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+brew install swiftlint
+```
+
+If you are unable to use Homebrew, you may use one of the other methods described in the [SwiftLint documentation](https://github.com/realm/SwiftLint).
```
Comment (ccjernigan) [inline SWIFTLINT.md:33]: Can we and a line for installing Homebrew before the bash command?

We don't have a link to the homebrew project, so at least pointing someone to it might be helpful context for a developer that's new to Apple platforms.

File: SWIFTLINT.md (hunk 23 above)
Comment (ccjernigan) [reply SWIFTLINT.md:30]: There's one small time bomb hiding here.  Because SwiftLint isn't managed as a versioned dependency, updates to the package can break our build over time.

This happened to me last year—they added a new check which was turned on by default and it failed my build on my CI server until I made a code change to correct the issue.

We avoid this issue on Android, because our linters are usually part of the Gradle build script and are therefore versioned.

I'm not sure there's a good solution to this problem... But at some point we'll probably hit this.  I suppose we could specify a version explicitly?  Or we could just add a note indicating this is a known possible issue?

File: SWIFTLINT.md [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+
+![](screens/swiftlint-warning.png)
+
+## Approved exceptions
+
+There are certain common idioms that violate SwiftLint's strict checking. If you are unable to work around them — and you've already tried to find a better way to structure your code — you may disable rules as described in this section.
+
+If you find that you're struggling with rules other than those described below, please reach out to your Team Lead with your specific example.
+
+### Implicitly Unwrapped Optionals
```
Comment (ccjernigan) [inline SWIFTLINT.md:85]: Minor text consistency: this is the only header that's using capitals for each word.

File: SWIFTLINT.md [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+
+```
+// swiftlint:disable all
+```
+
+## Other notes
+
+While SwiftLint goes a long way towards making your source code compliant with our style guide, it doesn't cover everything. For example, it won't catch or force you to correct the formatting for multi-condition `guard` statements. (See [Golden Path](https://github.com/raywenderlich/swift-style-guide#golden-path) for correct formatting.)
+
+This configuration has been tested against several dozen of our most recent tutorials. A couple of rules, such as the line length limit or the limit on the length of a function, may need tweaking to fit our style. If you find yourself butting heads with SwiftLint, please reach out to the iOS Team Lead with details.
```
Comment (ccjernigan) [inline SWIFTLINT.md:138]: > please reach out to the iOS Team Lead with details

How would an open source contributor reach out?

File: co.electriccoin.swiftlint.yml [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+# https://github.com/raywenderlich/swift-style-guide
+
+excluded:
+  - ${PWD}/Pods
+  - ${PWD}/xctemplates
+
+disabled_rules:
+  - notification_center_detachment
+  - orphaned_doc_comment
+  - todo
```
Comment (ccjernigan) [inline co.electriccoin.swiftlint.yml:11]: Do we want to disable todos?

I managed to configure a regex for detekt (a similar static analysis tool for Kotlin code), that allows todos if they have an issue number associated with them.  E.g. `TODO [#123]: This doesn't correctly support LTR languages` wouldn't trigger a failure but a plain `TODO: This doesn't correctly support LTR languages` would trigger a failure.

Does SwiftLint support anything similar?

File: secant/Screens/Welcome Screen/WelcomeScreenViewModel.swift [hunk 29]
```diff
@@ -10,6 +10,10 @@ import Combine
 
 class WelcomeScreenViewModel: BaseViewModel<Services>, ObservableObject {
     
+<<<<<<< Updated upstream
```
Comment (ccjernigan) [inline secant/Screens/Welcome Screen/WelcomeScreenViewModel.swift:13]: Looks like we have some git conflict markers here.

File: secantTests/OnboardingTests.swift [hunk 30]
```diff
@@ -0,0 +1,44 @@
+//
+//  OnboardingTests.swift
+//  secantTests
+//
+//  Created by Francisco Gindre on 9/8/21.
```
Comment (ccjernigan) [inline secantTests/OnboardingTests.swift:5]: What will be our convention on including authors?

File: secant.xcodeproj/project.pbxproj [hunk 31]
```diff
@@ -313,6 +317,12 @@
 			children = (
 				0D4E7A1A26B364180058B01E /* secantTests.swift */,
 				0D4E7A1C26B364180058B01E /* Info.plist */,
+<<<<<<< Updated upstream
```
Comment (ccjernigan) [inline secant.xcodeproj/project.pbxproj:320]: Looks like some git conflict markers here.

File: secant.xcodeproj/project.pbxproj [hunk 32]
```diff
@@ -548,7 +558,13 @@
 			isa = PBXSourcesBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+<<<<<<< Updated upstream
```
Comment (ccjernigan) [inline secant.xcodeproj/project.pbxproj:561]: Looks like some git conflict markers here.

File: .github/pull_request_template.md (hunk 19 above)
Comment (ccjernigan) [reply .github/pull_request_template.md:19]: I later discovered we have this under another markdown file, although a checklist item might also be worth considering.

File: .github/pull_request_template.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ccjernigan) [reply .github/pull_request_template.md]: I think we just need to tweak the instructions to include something like this: https://www.swiftbysundell.com/tips/gathering-test-coverage-in-xcode/

File: secant.xcodeproj/project.pbxproj (hunk 16 above)
Comment (pacu) [reply secant.xcodeproj/project.pbxproj:15]: ahggg,  you had one job `git`. Thanks for pointing this out

File: .github/pull_request_template.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (pacu) [reply .github/pull_request_template.md:8]: Yes, probably license changes won't be frequent

> I mention this, because reviewers might click the link and see the version on main but not realize that's what they're viewing
Yes, this is real. I don't know how to avoid this.

File: SWIFTLINT.md (hunk 23 above)
Comment (adamstener) [reply SWIFTLINT.md:30]: The script that executes SwiftLint in the project can contain a version check.

File: co.electriccoin.swiftlint.yml (hunk 3 above)
Comment (pacu) [reply co.electriccoin.swiftlint.yml:125]: adding
````
file_length:
  warning: 600
  ignore_comment_only_lines: false: true 
````

File: SWIFTLINT.md (hunk 27 above)
Comment (pacu) [reply SWIFTLINT.md:138]: I'll put my phone number :P 
Great catch

File: .github/pull_request_template.md (hunk 19 above)
Comment (pacu) [reply .github/pull_request_template.md:19]: Thanks for pointing this out. I need to find the manual test scripts.

File: co.electriccoin.swiftlint.yml (hunk 28 above)
Comment (pacu) [reply co.electriccoin.swiftlint.yml:11]: I don't know. probably this is possible by adding a custom rule as the ones @adamstener added

Comment (adamstener) [reply co.electriccoin.swiftlint.yml:11]: This would be custom yes but could likely use the same regex

File: SWIFTLINT.md (hunk 24 above)
Comment (ccjernigan) [reply SWIFTLINT.md:29]: I looked up the GitHub syntax highlighting [documentation](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml), and it all resolves to a "shell script" highlighter.  So either "bash" or "zsh" would have the same result here.

File: .github/pull_request_template.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (pacu) [reply .github/pull_request_template.md]: I'm adding a doc section to show how to check the code coverage %

File: .github/pull_request_template.md (hunk 19 above)
Comment (pacu) [reply .github/pull_request_template.md]: I added the placeholder section

**Review by ccjernigan — COMMENTED** [review]
This is really coming along—I just have a few minor non-blocking comments.

File: co.electriccoin.swiftlint.yml (hunk 28 above)
Comment (pacu) [reply co.electriccoin.swiftlint.yml:11]: probably that's for the best!

File: SWIFTLINT.md [hunk 33]
```diff
@@ -0,0 +1,136 @@
...
+```
+
+This violates the rule that you should not use trailing closure syntax when a method accepts multiple closure parameters, so SwiftLint will flag it with a warning. 
+
+You can work around this by extracting the `Button`'s action into a separate method. While this is frequently a better solution when the action requires several statements, it's an onerous requirement when the action is a single statement, as in the example above.
+
+In these cases, you're permitted to disable this rule **for the declaration of a SwiftUI view** only. The rule name is `multiple_closures_with_trailing_closure`.
+
+### Open-source files
+
```
[the remark this answers is not in the document]
Comment (pacu) [reply SWIFTLINT.md:127]: I removed this section. We haven't pulled files that way before.

File: co.electriccoin.swiftlint.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (pacu) [reply co.electriccoin.swiftlint.yml:144]: we can defer this to the integration of the linter to the actual code

[the remark this answers is not in the document]
Comment (adamstener) [reply co.electriccoin.swiftlint.yml:144]: Can you elaborate more on what you mean by that?

[the remark this answers is not in the document]
Comment (adamstener) [reply co.electriccoin.swiftlint.yml:144]: If max_empty_lines was 2, and it count comments as empty lines, then anywhere we add inline documentation or anything is going to have issues whether this is 1 or 2. I don't think that this rule considers comments as empty lines.

File: co.electriccoin.swiftlint.yml (hunk 28 above)
Comment (pacu) [reply co.electriccoin.swiftlint.yml:11]: done https://github.com/zcash/secant-ios-wallet/issues/25

File: co.electriccoin.swiftlint.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (pacu) [reply co.electriccoin.swiftlint.yml:144]: this can't be tested until the linter integrated to the project, ran and issues are solved.

File: CODE_REVIEW_GUIDELINES.md (hunk 22 above)
Comment (adamstener) [reply CODE_REVIEW_GUIDELINES.md:32]: I'm having a hard time trying to see any conflicting priorities here since this list is not overly concrete.  "well-designed" and "functionality is good" or "more complex than it needs to be" are all worded such that they aren't drawing hard lines that could cause conflicts.  Did you have anything specific in mind @ccjernigan ?

**Review by ccjernigan — APPROVED** [review]

File: CODE_REVIEW_GUIDELINES.md (hunk 22 above)
Comment (ccjernigan) [reply CODE_REVIEW_GUIDELINES.md:32]: Maybe I'm falling into the trap of `The developer isn’t implementing things they might need in the future but don’t know they need now or aren't using now.`

I imagine at some point we'll add security and performance to this list, both of which often have tradeoffs to consider.

@adamstener is right that the current items are phrased in a way that avoids conflict nicely!

Comment (adamstener) [reply CODE_REVIEW_GUIDELINES.md:32]: I imagine that this particular rule just needs an explanation to be broken is all.  If something needs to be implemented in an incomplete way, there should be a valid reason, and it probably won't conflict with anything else.

**Review by adamstener — APPROVED** [review]
Looks great!
